### PR TITLE
Creating new FPGA tutorial: Autorun

### PR DESCRIPTION
**Pull request by tyoungsc**
# Adding a New Sample(s)
## Description
This new FPGA tutorial shows how to create Autorun kernels. Autorun kernels are kernels that are submitted before main() starts and have no (direct) interaction with the host. This was a feature in OpenCL with special compiler/runtime support. The biggest reason to use autorun kernels was to avoid the logic for the kernel to communicate with the host. Today, this logic overhead is negligible, but OpenCL users will need to know how to convert their autorun kernels, which this tutorial helps with.

## Checklist
Administrative
- [X] Review sample design with the appropriate [Domain Expert](https://github.com/oneapi-src/oneAPI-samples/wiki/Reviewers-and-Domain-Experts): @mdbtucker 

Code Development
- [X] Implement coding guidelines and ensure code quality. [see wiki for details](https://github.com/oneapi-src/oneAPI-samples/wiki/General-Code-Guidelines)
- [X] Adhere to readme template 
- [X] Enforce format via clang-format config file
- [X] Adhere to sample.json specification. https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-specification
- [ ] Ensure/create CI test configurations for sample (ciTests field) https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-ci-test-object
- [ ] Run jsonlint on sample.json to verify json syntax. www.jsonlint.com

Security and Legal
- [ ] OSPDT Approval (see @JoeOster for assistance)
- [ ] Compile using the following compiler flags and fix any warnings, the falgs are: "/Wall -Wformat-security -Werror=format-security"
- [ ] Bandit Scans (Python only)
- [ ] Virus scan

Review
- [ ] Review DPC++ code with Paul Peterseon. (GitHub User: pmpeter1)
- [ ] Review readme with Tom Lenth(@tomlenth) and/or Joe Oster(@JoeOster)
- [ ] Tested using Dev Cloud when applicable

### CODE REVIEW

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/src/autorun.cpp
Comment (mdbtucker) [inline DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/src/autorun.cpp:8]: Should we put this into the shared include directory?

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/README.md
Comment (tyoungsc) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/README.md]: Deleted

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/README.md
Comment (tyoungsc) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/README.md:3]: Done

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/sample.json
Comment (tyoungsc) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/sample.json]: Forgot if I did or didn't, so regenerated.

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/src/autorun.cpp
Comment (tyoungsc) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/src/autorun.cpp:8]: Yes, but it is not ready yet and I want to ensure this tutorial gets in.

Comment (mdbtucker) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/src/autorun.cpp:8]: It's ready to be comitted for 2022.2 as soon as the 2022.1 label is created (early next week).  See https://github.com/oneapi-src/oneAPI-samples/pull/680.  It will be committed before this new sample is committed.

Some options for you:
1. Update this sample now to use the common include directory.
2. Wait until the common include directory PR is committed, then update this sample before we merge it.
3. We merge this sample as is, and sometime in 2022.2 you update this sample to move the header to the common include directory.

Your choice, just let me know which you plan to do.

**Review by mdbtucker — APPROVED** [review]

**Review by tomlenth — APPROVED** [review]
looks good, thx

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/src/autorun.hpp
Comment (keryell) [inline DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/src/autorun.hpp:90]: Missing EOL.

Comment (tyoungsc) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/src/autorun.hpp:90]: Here is the thing. I was always told to have a newline at the end of files. I forget why, but I was always told to. However, clang-format doesn't do it for me and I don't get a warning. I remember in a previous tutorial I had newlines at the end of files and either someone or clang-format made me get rid of them.

Anyways. Long rant for me to say I am not sure what to do. @mdbtucker  ??

Comment (mdbtucker) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/src/autorun.hpp:90]: If clang-format doesn't add the eol, then I don't think you NEED to add it.  The EOL doesn't seem to be part of the Google coding style guide, and I'm not aware of any guidelines in the Code Samples that says we have to do this.

That said, if clang-format doesn't REMOVE the eol, then I think it's at your discretion.  The argument I've found online for why to do this is if you add new lines to the end of the file in a future edit and you don't have an EOL at the end of the file, then the change will show the last line of the file as being edited (to add the EOL).  I think this is pretty weak justification, but it is a justification I guess.

Long story short, I really don't see a good reason to care one way or the other.  I wouldn't fail to approve a PR because of this, but I would also approve a future PR if somebody who does care wants to go in and add that EOL.

**Comment by tyoungsc** [comment]
@tomlenth @mdbtucker I changed the base branch from development to master, which dismissed your reviews.

**Review by mdbtucker — DISMISSED** [review]
Re-approving after base branch was updated from development to master.

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/src/autorun.hpp
Comment (mdbtucker) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/src/autorun.hpp:90]: Nobody seems to care too much about this one, marking it as resolved.  @keryell feel free to reopen if you disagree.

**Comment by mdbtucker** [comment]
@tyoungsc any idea why the Windows emulation run failed in the CI?  Can you investigate?  Maybe add a few debug prints and 'push' to this branch (which will trigger the CI to run again)?

**Comment by tyoungsc** [comment]
> @tyoungsc any idea why the Windows emulation run failed in the CI? Can you investigate? Maybe add a few debug prints and 'push' to this branch (which will trigger the CI to run again)?

Does the CI infra treat all warnings as errors? If so, the failure is likely because of the warning shown in the log.

**Comment by mdbtucker** [comment]
> @mdbtucker Does the CI infra treat all warnings as errors? If so, the failure is likely because of the warning shown in the log.

No it does not.  All the other samples have this same warnings under Windows.

**Comment by tyoungsc** [comment]
@mdbtucker can you re approve after my changes to start CI/CD again please :)?

**Review by mdbtucker — DISMISSED** [review]

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/src/autorun.hpp
Comment (keryell) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/src/autorun.hpp:90]: At the end of the day, I do not care that much about improving the code of my competitors. :-)
So you can even remove the EOL at the end of all the files to make your code quality worse, even it is sad to have bad code quality on SYCL examples for FPGA in general... :-( @bader

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/src/autorun.cpp
Comment (tyoungsc) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/src/autorun.cpp:8]: 3 is my choice.

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/src/autorun.hpp
Comment (tyoungsc) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/src/autorun.hpp:90]: We are following the Google coding style for clang-format, which doesn't add newlines to the end of files even though clang-format presents that option.

Comment (tyoungsc) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/autorun/src/autorun.hpp:90]: Moreover, based on the C++11 spec (I think this started in C++03), this is no longer necessary with a complian compiler:

C++03 Standard [2.2.1.3] ([page 31](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2013/n3690.pdf))

> A source file that is not empty and that does not end in a new-line character, or that ends in a new-line character
immediately preceded by a backslash character before any such splicing takes place, shall be processed
as if an additional new-line character were appended to the file.

I think the only remaining argument is how files that don't do this "play" with other tools (e.g. VIM, cat, ...)? I concede that these other tools may not like this, but we don't think this makes it worth deviating from the Google coding standards for; those people seem to know what they're doing.

**Comment by mdbtucker** [comment]
This PR is currently stuck waiting on a resolution to https://jira.devtools.intel.com/projects/REIE/issues/REIE-1225 and https://jira.devtools.intel.com/projects/REIE/issues/REIE-1224.

**Review by mdbtucker — APPROVED** [review]
Approving to trigger CI after adding a workaround to the Windows Emulator hangs (see https://jira.devtools.intel.com/browse/CMPLRLLVM-34145)

**Review by mdbtucker — APPROVED** [review]

**Pull request merged**
